### ISSUE TITLE
Bug: Subscribing multiple endpoints to the same Topic not longer supported

### ISSUE THREAD

**Issue by fbridger** [dapr/dotnet-sdk#752]
Does the Dapr DotNet-SDK supports subscribing multiple endpoints to the same Topic?

I was under the assumption that this basic scenario was obviously supported but I can't seem to make it work. Seems extremely weird that we can't subscribe multiple/different endpoints (even in different Controllers) to the same Topic.

Here is the branch with all the code mentioned below: https://github.com/fbridger/dotnet-sdk/tree/pubsub-multiple-subscribers

# Trying to subscribe to the same Topic

Example being modified: https://github.com/dapr/dotnet-sdk/tree/master/examples/AspNetCore/ControllerSample

In the [ControllerSample (included in dotnet-sdk)](https://github.com/dapr/dotnet-sdk/tree/master/examples/AspNetCore/ControllerSample), I am trying to subscribe multiple endpoints to same Topic but can't seem to make it work.

I've added the following methods in the `SampleController.cs`
```csharp
[Topic("pubsub", "deposit")]
[HttpPost("LogDeposit")]
public ActionResult LogDeposit(Transaction transaction)
{
    logger.LogDebug($"Log deposit Id: {transaction.Id} Amount: {transaction.Amount}");
    return NoContent();
}

[Topic("pubsub", "withdraw")]
[HttpPost("LogWithdraw")]
public ActionResult LogWithdraw(Transaction transaction)
{
    logger.LogDebug($"Log withdraw Id: {transaction.Id} Amount: {transaction.Amount}");
    return NoContent();
}
```

However, when running the app and hitting the /dapr/subscribe those new endpoints are not registered:
```json
[
  {
    "topic": "custom-topic",
    "pubsubName": "custom-pubsub",
    "route": "exampleCustomTopic"
  },
  {
    "topic": "deposit",
    "pubsubName": "pubsub",
    "route": "deposit"
  },
  {
    "topic": "withdraw",
    "pubsubName": "pubsub",
    "routes": {
      "default": "withdraw",
      "rules": [
        {
          "match": "event.type ==\"withdraw.v2\"",
          "path": "withdraw.v2"
        }
      ]
    }
  }
```

Applications logs:
```
== APP == fail: DaprTopicSubscription[0]
== APP ==       A default subscription to topic deposit on pubsub pubsub already exists.
== APP == fail: DaprTopicSubscription[0]
== APP ==       A default subscription to topic withdraw on pubsub pubsub already exists.
== APP == fail: DaprTopicSubscription[0]
== APP ==       A default subscription to topic custom-topic on pubsub custom-pubsub already exists.
```

**Comment by fbridger** [comment]
@pkedy I have confirmed that after the changes you applied in [aba49d1446dcd5fa68aa5b8a7597cb6a59e83f88](https://github.com/dapr/dotnet-sdk/commit/aba49d1446dcd5fa68aa5b8a7597cb6a59e83f88) the /dapr/subscribe is removing Topics with the same Topic and Name but different Route.

Here is the branch where I've reverted your changes and added multiple endpoints to subscribe to the same Topic: https://github.com/fbridger/dotnet-sdk/tree/pubsub-multiple-subscribers-before-changes

I've reverted your changes and the /dapr/subscribe endpoint respond with the expected result:
```json
// 20210911111320
// http://localhost:5000/dapr/subscribe

[
  {
    "topic": "deposit",
    "pubsubName": "pubsub",
    "route": "deposit"
  },
  {
    "topic": "withdraw",
    "pubsubName": "pubsub",
    "route": "withdraw"
  },
  {
    "topic": "custom-topic",
    "pubsubName": "custom-pubsub",
    "route": "exampleCustomTopic"
  },
  {
    "topic": "deposit",
    "pubsubName": "pubsub",
    "route": "LogDeposit"
  },
  {
    "topic": "withdraw",
    "pubsubName": "pubsub",
    "route": "LogWithdraw"
  }
]
```

**Comment by fbridger** [comment]
I've added a Test to be able to reproduce this error:
https://github.com/fbridger/dotnet-sdk/commit/7194de11c76c5c0caac07731e1092f039f9fcb02

**Comment by rynowak** [comment]
@fbridger - thanks for filing this, we'll take a look.

**Comment by halspang** [comment]
@fbridger - What is the behavior that you're trying to achieve here and did it work on an older version of the .NET SDK or dapr runtime?

My understanding of the Dapr runtime is that it will actually eat (silently, which is bad) the duplicate endpoints, so I don't think that both endpoints are being called. I took the pubsub example and tried to get that to happen, here's the code and what dapr did:

Node.js change:
```node
app.get('/dapr/subscribe', (_req, res) => {
    res.json([
        {
            pubsubname: "pubsub",
            topic: "A",
            route: "A"
        },
        {
            pubsubname: "pubsub",
            topic: "B",
            route: "B"
        },
        {
            pubsubname: "pubsub",
            topic: "A",
            route: "B"
        }
    ]);
});
```

Dapr:
```
DEBU[0000] try to connect to placement service: dns:///localhost:50005  app_id=node-subscriber instance=SATSOWPC.fareast.corp.microsoft.com scope=dapr.runtime.actor.internal.placement type=log ver=1.3.0
DEBU[0000] app responded with subscriptions [{pubsub A A map[] []} {pubsub B B map[] []} {pubsub A B map[] []}]  app_id=node-subscriber instance=SATSOWPC.fareast.corp.microsoft.com scope=dapr.runtime type=log ver=1.3.0
INFO[0000] app is subscribed to the following topics: [B A] through pubsub=pubsub  app_id=node-subscriber instance=SATSOWPC.fareast.corp.microsoft.com scope=dapr.runtime type=log ver=1.3.0
DEBU[0000] subscribing to topic=A on pubsub=pubsub       app_id=node-subscriber instance=SATSOWPC.fareast.corp.microsoft.com scope=dapr.runtime type=log ver=1.3.0
DEBU[0000] subscribing to topic=B on pubsub=pubsub       app_id=node-subscriber instance=SATSOWPC.fareast.corp.microsoft.com scope=dapr.runtime type=log ver=1.3.0
INFO[0000] dapr initialized. Status: Running. Init Elapsed 242.696ms  app_id=node-subscriber instance=SATSOWPC.fareast.corp.microsoft.com scope=dapr.runtime type=log ver=1.3.0
```

You can see that Dapr sees that my app responded with the 3 endpoints, but then only subscribes to 2 of them. It looks like this section of code does the removal:

https://github.com/dapr/dapr/blob/master/pkg/runtime/runtime.go#L1179

**Comment by fbridger** [comment]
@halspang I am trying to subscribe to the same Topic (e.g. "B") multiple times from the same Application (e.g. routes "B1" and "B2").

Previous to the PR #746 the dotnet-sdk was generating the correct json when hitting "/dapr/subscribe"

/dapr/subscribe JSON **before** the PR:
```
[
{
    pubsubname: "pubsub",
    topic: "B",
    route: "B1"
},
{
    pubsubname: "pubsub",
    topic: "B",
    route: "B2"
}
]
```

/dapr/subscribe JSON **after** the PR:
```
[
{
    pubsubname: "pubsub",
    topic: "B",
    route: "B1"
}
]
```

Route B2 is now being excluded so, an Application can NO LONGER subscribe multiple times to the same Topic.

**Comment by rynowak** [comment]
@fbridger - what we're trying to understand is whether this was ever working E2E. I understand that the .NET SDK does something different. From looking at the runtime code this had no effect because the duplicate was discarded by the runtime.

**Comment by fbridger** [comment]
I believe your E2E was not covering the scenario that I a mentioning. Which is currently supported by the Dapr SideCar.

As I mentioned in my previous comment, if you checkout the commit previous to the one that @pkedy made. You will get the dotnet-sdk to correctly include multiple routes using the same Topic.

Previously, you could subscribe to the same Topic from 2 different operations (routes).

**Comment by halspang** [comment]
@fbridger - I understand that the JSON changed, but what I'm seeing in the code appears to show that the sidecar does not support this. We've seen that *separate* applications are able to subscribe to the same topic and if the underlying pubsub supports it, it can work correctly. 

With my sample though, you can see that even though the application returns 3 endpoints, the sidecar only subscribes 2 because 1 of them is overwritten.

**Comment by rynowak** [comment]
@fbridger - can you share the version of the Dapr runtime and which pubsub component you are using so we can try to investigate this properly? We don't intend to cut off things that were working, but I think we need to convince ourselves that this was possible.

**Comment by fbridger** [comment]
@halspang @rynowak  You are both correct. 

It seems that the Dapr SideCar does not support the same Application to subscribe multiple Times to the same Topic.

I'll request the addition of this feature to the Dapr Dapr team.

I am very sorry for the time you spent looking into this issue and really appreaciate your comments.

Thanks!

**Comment by fbridger** [comment]
Requested feature in Dapr Dapr: https://github.com/dapr/dapr/issues/3664

**Comment by rynowak** [comment]
@fbridger - thanks! No worries, we're here to help.

**Comment by georgestevens99** [comment]
FYI A long, long time ago, before 1.0.0 was released, I did a test using the Dapr .NET SDK to verify that pubsub using both redis and Azure Service Bus Topics did indeed support multiple subscribers.  I clearly remember this because I had to add a 3rd service (ServiceC) to my 2 service (ServiceA and ServiceB) test system in order to test multiple subscribers. ServiceA is the publisher.

Also keep in mind that the essence of a Topic (aka PubSub) is its ability to publish a single event to multiple subscribers.  This is in contrast to a Queue which can publish a message to one, and only one, listener.  So Dapr definitely needs to repair this to support the essence of a PubSub.

Thanks for your collective work.
